From 93b5e214d6ababfacbd108b6aa717f4ad1f07166 Mon Sep 17 00:00:00 2001
From: Benjamin Coddington <bcodding@redhat.com>
Date: Thu, 8 Jan 2015 21:43:18 -0500
Subject: [fs] LOCKD: Fix a race when initialising nlmsvc_timeout

Message-id: <89cde1bbd7afcfe755979ca4323c097a59c040f7.1420752997.git.bcodding@redhat.com>
Patchwork-id: 102551
O-Subject: [RHEL7.1 PATCH fs BZ1144982] LOCKD: Fix a race when initialising nlmsvc_timeout
Bugzilla: 1144982
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

This commit fixes a race whereby nlmclnt_init() first starts the lockd
daemon, and then calls nlm_bind_host() with the expectation that
nlmsvc_timeout has already been initialised. Unfortunately, there is no
no synchronisation between lockd() and lockd_up() to guarantee that this
is the case.

Fix is to move the initialisation of nlmsvc_timeout into lockd_create_svc

Fixes: 9a1b6bf818e74 ("LOCKD: Don't call utsname()->nodename...")
Cc: Bruce Fields <bfields@fieldses.org>
Cc: stable@vger.kernel.org # 3.10.x
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 06bed7d18c2c07b3e3eeadf4bd357f6e806618cc
Bugzilla: 1144982

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/lockd/svc.c b/fs/lockd/svc.c
index d56a990..7193792 100644
--- a/fs/lockd/svc.c
+++ b/fs/lockd/svc.c
@@ -137,10 +137,6 @@ lockd(void *vrqstp)
 
  dprintk("NFS locking service started (ver " LOCKD_VERSION ").\n");
 
- if (!nlm_timeout)
-  nlm_timeout = LOCKD_DFLT_TIMEO;
- nlmsvc_timeout = nlm_timeout * HZ;
-
  /*
   * The main request loop. We don't terminate until the last
   * NFS mount or NFS daemon has gone away.
@@ -348,6 +344,10 @@ static struct svc_serv *lockd_create_svc(void)
   printk(KERN_WARNING
    "lockd_up: no pid, %d users??\n", nlmsvc_users);
 
+ if (!nlm_timeout)
+  nlm_timeout = LOCKD_DFLT_TIMEO;
+ nlmsvc_timeout = nlm_timeout * HZ;
+
  serv = svc_create(&nlmsvc_program, LOCKD_BUFSIZE, NULL);
  if (!serv) {
   printk(KERN_WARNING "lockd_up: create service failed\n");
-- 
1.7.1